server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueFactory.java

```diff
@@ -0,0 +1,92 @@
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.queue.rabbitmq;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.apache.james.queue.api.MailQueue;
+import org.apache.james.queue.api.MailQueueFactory;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+
+
+public class RabbitMQMailQueueFactory implements MailQueueFactory<RabbitMQMailQueue> {
+
+    private static final String ROUTING_KEY = "";
+    private static final boolean RABBIT_OPTION_DURABLE = true;
+    private static final boolean RABBIT_OPTION_EXCLUSIVE = true;
+    private static final boolean RABBIT_OPTION_AUTO_DELETE = true;
+
+    private final Channel channel;
+    private final RabbitMQManagementApi mqManagementApi;
+
+    @VisibleForTesting
+    @Inject
+    RabbitMQMailQueueFactory(Connection connection, RabbitMQManagementApi mqManagementApi) throws IOException {
+        this.channel = connection.createChannel();
+        this.mqManagementApi = mqManagementApi;
+    }
+
+    @Override
+    public Optional<RabbitMQMailQueue> getQueue(String name) {
+        return getQueue(MailQueueName.fromString(name));
+    }
+
+    @Override
+    public RabbitMQMailQueue createQueue(String name) {
+        MailQueueName mailQueueName = MailQueueName.fromString(name);
+        return getQueue(mailQueueName)
+            .orElseGet(() -> attemptQueueCreation(mailQueueName));
+    }
+
+    @Override
+    public Set<RabbitMQMailQueue> listCreatedMailQueues() {
+        return mqManagementApi.listCreatedMailQueueNames()
+            .map(RabbitMQMailQueue::new)
+            .collect(Guavate.toImmutableSet());
+    }
+
+    private Optional<RabbitMQMailQueue> getQueue(MailQueueName name) {
+        return mqManagementApi.listCreatedMailQueueNames()
+            .filter(name::equals)
+            .map(RabbitMQMailQueue::new)
+            .findFirst();
```

Comment:
Conceptually I prefer having `findFirst`then `map`, it's more straightforward.

Reply:
really ? I think the exact opposite

Reply:
I agree that in the declarative style the result is the same, but when I read it I think the following:

 * `mqManagementApi.listCreatedMailQueueNames()' => I get all the Queue names
 * `.filter(name::equals)` => I keep only those matching
 * `.map(RabbitMQMailQueue::new)` => I strong typed all the matching names
 * `.findFirst()` => I keep only the first one

Reply:
Yes, it's (almost) what it does and as you know, the `.map(RabbitMQMailQueue::new)` is lazily executed so we build just one object. Lazy or not, the result is the same, one is faster then the other one.
One reason I prefer my version is the ability to compose : we can use a filtered strong-typed stream in several place and use `findFirst` only in some cases.
If you put the `findFirst` in the middle of the stream, you can't refactor it easily.
Also, it's a kind of terminal operation (it triggers things on the stream) so I usually split it from stream "definition" (what will happen when we will evaluate the stream).